### PR TITLE
Issue #3892: Added TokenTypes.STATIC_IMPORT in NoLineWrapCheck

### PR DESCRIPTION
**Pull request by kukreja-vikramaditya**
Issue: #3892

### CODE REVIEW

**Comment by MEZk** [comment]
@kukreja-vikramaditya
As you found the problem and fixed it by adding a new token, I think it is worth adding some very simple example into xdoc.

**Comment by kukreja-vikramaditya** [comment]
@romani Will add UTs as suggested
@MEZk There is an issue #3625 which requests for more examples of NoLineWrapCheck in xdocs. Should I make the changes for that issue here itself or in a new PR?

**Comment by MEZk** [comment]
@kukreja-vikramaditya
Examples that you did for this issue are OK . Thanks.

**Comment by kukreja-vikramaditya** [comment]
@MEZk I am getting error
```xml
Import statement for 'java.math.BigInteger.ZERO' is in the wrong order. Should be in the 'STATIC' group, expecting not assigned imports on this line. [CustomImportOrder]
```

But as per the checkstyle documentation [here](http://checkstyle.sourceforge.net/config_imports.html#ImportOrder) for IDEA, my code should do just fine

```java
package com.google.checkstyle.test.chapter3filestructure.rule332nolinewrap; //ok

import com.puppycrawl.tools.checkstyle.checks.design.FinalClassCheck; //ok
 
import java.util.concurrent.atomic.AtomicIntegerFieldUpdater; //ok
import javax.accessibility.AccessibleAttributeSequence; //ok

import static java.math.BigInteger.ZERO; //ok

public class InputNoLineWrapGood {
    
  public void fooMethod() {
      //
  }
}
```

**Comment by kukreja-vikramaditya** [comment]
PR passed all tests and UTs and examples were updated too

**Comment by kukreja-vikramaditya** [comment]
@rnveach Will add the changes and update soon.

**Comment by kukreja-vikramaditya** [comment]
@rnveach Could you please review Issue #3932 also and let me know how to proceed

File: src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapBad.java [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
-import com.puppycrawl.tools.checkstyle.checks.design.FinalClassCheck; //ok
+import static java.math. //warn
+        BigInteger.ZERO;
 
-import javax.accessibility. //warn
-    AccessibleAttributeSequence;
+import com.puppycrawl.tools.checkstyle.checks.design.FinalClassCheck; //ok
 import java.util.concurrent.atomic.AtomicIntegerFieldUpdater; //ok
+import javax.accessibility. //warn
+        AccessibleAttributeSequence;
```
Comment (romani) [inline src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapBad.java]: please make a changes to this file as min as possible. You add two lines and all other are changed for now reason.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
-import com.puppycrawl.tools.checkstyle.TreeWalker;
+import static java.math.
+	BigInteger.ZERO;
 
-import javax.accessibility.
-    AccessibleAttributeSequence;
+import com.puppycrawl.tools.checkstyle.TreeWalker;
 import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import javax.accessibility.
+		AccessibleAttributeSequence;
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java]: please make a changes to this file as min as possible. You add two lines and all other are changed for now reason.

Please do the same for all Input files your touched.

File: src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapBad.java (hunk 1 above)
Comment (kukreja-vikramaditya) [inline src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapBad.java]: I'll make the changes as you suggested but is it ok to defy the import order as per google's java guidelines [here](https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing) even if it is "Bad" example? I reordered the imports as given in the link. Still no worries I'll make the changes as you said

**Comment by kukreja-vikramaditya** [comment]
@rnveach I didn't understand as to why extra tests would be required? I have the same three Input files (of which I only touched two)... Only the line number needs to be updated in the "Bad" example Test right?

**Comment by rnveach** [comment]
> I didn't understand as to why extra tests would be required?

See http://checkstyle.sourceforge.net/google_style.html .
We reference tests directly from there to showcase examples specifically to people using the google configuration. It is a lot easier than trying to pinpoint specific tests in our full testing area.
So now anyone using the google configuration can go and see specific examples we check for.

File: src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapBad.java (hunk 1 above)
Comment (romani) [inline src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapBad.java]: this is "Input" file , code in such files are suppose to be ugly as it is possible. Style guides are not applicable to them at all.
shifting of line numbers is ok, if shifting is affect a lot of code it is recommended to create new Input  file and put most new cases in it.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java (hunk 2 above)
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java]: @kukreja-vikramaditya , still not addressed .

Comment (kukreja-vikramaditya) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java]: @romani There is an issue I face everytime  the order of imports in `InputNoLineWrapBad` is changed. The travis ci fails at Subtask No. 16, which says that static should be in its own group
What i do to pass the test
```java
package com.puppycrawl.tools.
    checkstyle.checks.whitespace;

import static java.math.
	BigInteger.ZERO;

import com.puppycrawl.tools.checkstyle.TreeWalker;
import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
import javax.accessibility.
		AccessibleAttributeSequence;

public class
    InputNoLineWrapBad {

	public void
	    fooMethod() {
		final int
		    foo = 0;
	}
}

enum
    FooFoo {
}

interface
    InterFoo {}
```
Where the imports are in placed as mentioned here - https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing

But this what you saying i shouldn't do...Please help in this regard

Comment (kukreja-vikramaditya) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java]: I can push the branch again if you wish to see the failed CI subtask

Comment (MEZk) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java]: @kukreja-vikramaditya 
I do not your problem to understand too. Checkstyle does not validate test input files.
In your case, for example, for InputNoLineWrapGood this works fine:
```java
import static java.math.BigInteger.ZERO;

import com.puppycrawl.tools.checkstyle.TreeWalker;

import javax.accessibility.AccessibleAttributeSequence;
import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;

public class InputNoLineWrapGood {
```

As you can see, I just added new static import and did not touch any other imports in the file.

**Comment by romani** [comment]
@kukreja-vikramaditya , please follow recommendations/demands from us. Tests are always not enough. The more tests the better.

In addition is is required to provide 2 diff reports to prove that there is no regression in Check:
- diff report without new token (between old and new version of code).
- diff report it a diff between previous version of Check and new version of Check with new token activated.

**Comment by rnveach** [comment]
@kukreja-vikramaditya Everything ok here?

**Comment by romani** [comment]
@kukreja-vikramaditya , ping.

**Comment by kukreja-vikramaditya** [comment]
I couldn't continue because I didn't understand the diff concept. No worries now. I'll create the reports for this PR after I update diff reports for PR #3928  in the next few hours

**Comment by MEZk** [comment]
@kukreja-vikramaditya 
You should generate 2 diff reports:
1) Checkstyle's config which is used by diff.groovy to generate diff reports should be the same for master branch and for the branch which contains your changes. The config should not add STATIC_IMPORT token into a set of tokens which are analyzed by the check. The idea is to prove that your changes will not introduce regression to existing functionality.
2) Checkstyle's config which is used by diff.groovy to generate diff should be different to the config for the branch which contains your changes. The config for the master branch should not have STATIC_IMPORT token in the set of tokens which are analyzed by the check, but the config for your branch should include the token. The idea of diff report generation is to prove that STATIC_IMPORT is really analyzed by the check and your changes does make sense. 

diff.groovy allows you to specify two Checkstyle's configuration files for the second purpose.

**Comment by kukreja-vikramaditya** [comment]
@MEZk Thanks a lot for the detailed explanation! I am currently downloading the different repositories as mentioned in the .properties files and will soon update the diff reports along with the minor change suggested by @romani

File: src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapBad.java (hunk 1 above)
Comment (kukreja-vikramaditya) [inline src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapBad.java]: > please make a changes to this file as min as possible. You add two lines and all other are changed for no reason.

By changed do you mean all the lines after that are shifted by 2 or I made some other change i shouldn't have?

Also, regarding  @rnveach review, i need to add more tests in the `it` folder. I was thinking of something like 
1. renaming badLineWrapTest to `defaultBadLineWrapTest` 
 and goodLineWrapTest to `defaultGoodLineWrapTest`
2. creating new tests called `ignoreStaticImportBadLineWrapTest` and `ignoreStaticImportGoodLineWrapTest`, but using the `Configuration` class's object, I can't edit the attributes. Is it ok if I use the `DefaultConfiguration` class instead?

**Comment by romani** [comment]
@kukreja-vikramaditya , please do some minor changes in commit to reset PR and relaunch all CIs, wercker had network problem.

**Comment by kukreja-vikramaditya** [comment]
Werker failed again.

**Comment by MEZk** [comment]
@kukreja-vikramaditya 
Reports look OK.
Werker failed due to connection problem.
```
Cloning into 'hipparchus'...
fatal: unable to access 'https://github.com/Hipparchus-Math/hipparchus.git/': Could not resolve host: github.com
```

**Comment by MEZk** [comment]
@kukreja-vikramaditya 
Please, do an empty push to trigger wercker as it is not able to get your commit.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java (hunk 2 above)
Comment (kukreja-vikramaditya) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java]: Done

File: src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapGood.java [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
 import javax.accessibility.AccessibleAttributeSequence; //ok
 
 public class InputNoLineWrapGood {
-    
+
   public void fooMethod() {
       //
   }
-}
+}
```
Comment (rnveach) [inline src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapGood.java]: Please revert this change. Files should end with atleast 1 new line.

File: src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapGood.java [hunk 4]
```diff
@@ -1,12 +1,14 @@
 package com.google.checkstyle.test.chapter3filestructure.rule332nolinewrap; //ok
 
+import static java.math.BigInteger.ZERO;
```
Comment (romani) [inline src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapGood.java]: please move this import to the bottom to minimize changes in UT files.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java [hunk 5]
```diff
@@ -1,6 +1,9 @@
 package com.puppycrawl.tools.
     checkstyle.checks.whitespace;
 
+import static java.math.
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java]: please move this import to the bottom to minimize changes in UT files.

please add comment " //violation" on line where you expect it.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapGood.java [hunk 6]
```diff
@@ -1,12 +1,14 @@
 package com.puppycrawl.tools.checkstyle.checks.whitespace;
 
+import static java.math.BigInteger.ZERO;
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapGood.java]: please move this import to the bottom to minimize changes in UT files.

File: src/xdocs/config_whitespace.xml [hunk 7]
```diff
@@ -843,6 +847,9 @@ package com.puppycrawl.
 
 import com.puppycrawl.tools.
     checkstyle.api.AbstractCheck;
+
+import static java.math.
```
Comment (romani) [inline src/xdocs/config_whitespace.xml]: please add comment " //violation" on line where you expect it. for other cases too in this example.

**Comment by kukreja-vikramaditya** [comment]
@romani
 At https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoLineWrapBad.java#L24 there is an interface which is line wrapped but inteface token is not added in the test file . Was this purposely done or accidently skipped? Do you want me to add it before pushing the new changes?

**Comment by romani** [comment]
> Was this purposely done or accidently skipped?

any code is ok in inputs, it just show that Check do not care of interfaces in this case/configuration.

File: src/xdocs/config_whitespace.xml (hunk 7 above)
Comment (kukreja-vikramaditya) [inline src/xdocs/config_whitespace.xml]: Should I add `//ok` also for good cases or do that in #3625 ?

File: src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapGood.java (hunk 4 above)
Comment (kukreja-vikramaditya) [inline src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapGood.java]: I shifted this after all the other import statements and i have this job failing
https://travis-ci.org/checkstyle/checkstyle/jobs/212226897

This task fails and the reason mentioned is 
```bash
[WARN] /home/travis/build/checkstyle/checkstyle/src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapGood.java:8: Import statement for 'java.math.BigInteger.ZERO' is in the wrong order. Should be in the 'STATIC' group, expecting not assigned imports on this line. [CustomImportOrder]
```

How should I proceed?

File: src/xdocs/config_whitespace.xml (hunk 7 above)
Comment (romani) [inline src/xdocs/config_whitespace.xml]: no need

File: src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapGood.java (hunk 4 above)
Comment (romani) [inline src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule332nolinewrap/InputNoLineWrapGood.java]: your a lucky - https://github.com/checkstyle/checkstyle/blob/master/.ci/travis/travis.sh#L22

as by its name "InputNoLineWrapGood.java"
this file should be correct (no violations) - so please put it to proper place, no changes in UTs expected, as no violations should be.
I it was my bad request to places it in wrong place.

**Comment by kukreja-vikramaditya** [comment]
@romani, @rnveach 
All changes are done. Please review

**Comment by romani** [comment]
@rnveach , please do final review
